app/code/Magento/CatalogUrlRewrite/Model/ProductScopeRewriteGenerator.php

```diff
@@ -143,13 +147,16 @@ public function generateForGlobalScope($productCategories, Product $product, $ro
         $productId = $product->getEntityId();
         $mergeDataProvider = clone $this->mergeDataProviderPrototype;
 
+        $changedStoreIds = $this->isCategoryRewritesEnabled()
+            ? $this->getChangedCategoriesStoreIds($product)
+            : [];
         foreach ($product->getStoreIds() as $id) {
-            if (!$this->isGlobalScope($id) &&
-                !$this->storeViewService->doesEntityHaveOverriddenUrlKeyForStore(
+            if ((!empty($changedStoreIds) && in_array((int)$id, $changedStoreIds))
```

Comment:
Could be simplified I believe to
```suggestion
            if (in_array((int)$id, $changedStoreIds)
```